File: python/2461.maximum-sum-of-distinct-subarrays-with-length-k.py
from collections import defaultdict


class Solution:
    def maximumSubarraySum(self, nums: list[int], k: int) -> int:
        maxSubSum = 0

        cur = defaultdict(int)
        curSum = 0

        for i, num in enumerate(nums):
            cur[num] += 1
            curSum += num

            if i >= k:
                cur[nums[i - k]] -= 1
                curSum -= nums[i - k]

                if cur[nums[i - k]] == 0:
                    del cur[nums[i - k]]

            if i >= k - 1 and len(cur) == k:
                maxSubSum = max(maxSubSum, curSum)

        return maxSubSum

File: cpp/2461.maximum-sum-of-distinct-subarrays-with-length-k.cpp
#include <unordered_map>
#include <vector>

using std::vector, std::unordered_map;

#include <iostream>
const int ZERO = []() {
  std::ios_base::sync_with_stdio(false);
  std::cin.tie(nullptr);
  return 0;
}();

class Solution {
public:
  long long maximumSubarraySum(vector<int> &nums, int k) {
    long maxSubSum = 0L;
    long curSum = 0L;
    unordered_map<int, int> window;

    for (int i = 0; i < nums.size(); i++) {
      curSum += nums[i];
      window[nums[i]]++;

      if (i >= k) {
        curSum -= nums[i - k];
        window[nums[i - k]]--;

        if (window[nums[i - k]] == 0) {
          window.erase(nums[i - k]);
        }
      }

      if (i >= k - 1 && window.size() == k) {
        maxSubSum = std::max(maxSubSum, curSum);
      }
    }

    return maxSubSum;
  }
};